node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt

```
@@ -453,65 +536,74 @@ class NodeLoginModule : LoginModule {
+            val nameCallback = NameCallback("Username: ")
+            val passwordCallback = PasswordCallback("Password: ", false)
+            val certificateCallback = CertificateCallback()
+
+            try {
+                callbackHandler.handle(arrayOf(nameCallback, passwordCallback, certificateCallback))
+            } catch (e: IOException) {
+                throw LoginException(e.message)
+            } catch (e: UnsupportedCallbackException) {
+                throw LoginException("${e.message} not available to obtain information from user")
+            }
 
-        log.info("Processing login for $username")
+            val username = nameCallback.name ?: throw FailedLoginException("Username not provided")
+            val password = String(passwordCallback.password ?: throw FailedLoginException("Password not provided"))
 
-        val validatedUser = if (username == PEER_USER || username == NODE_USER) {
+            log.info("Processing login for $username")
             val certificates = certificateCallback.certificates ?: throw FailedLoginException("No TLS?")
-            authenticateNode(certificates, username)
-        } else {
-            // Otherwise assume they're an RPC user
-            authenticateRpcUser(password, username)
+            val validatedUser = when (determineUserRole(certificates, username)) {
+                PEER_ROLE -> authenticatePeer(certificates)
+                NODE_ROLE -> authenticateNode(certificates)
+                VERIFIER_ROLE -> authenticateVerifier(certificates)
+                RPC_ROLE -> authenticateRpcUser(password, username)
+                else -> throw FailedLoginException("Peer does not belong on our network")
+            }
+            principals += UserPrincipal(validatedUser)
+
+            loginSucceeded = true
+            return loginSucceeded
+        } catch (exception: FailedLoginException) {
```

I added this because these exceptions are completely swallowed otherwise. Should I rethrow?